src/vanilla/store.ts

```diff
@@ -487,42 +485,72 @@ export const createStore = () => {
       })
       return dependents
     }
-    const loop1 = (a: AnyAtom) => {
-      getDependents(a).forEach((dependent) => {
-        if (dependent !== a) {
-          dependencyMap.set(
-            dependent,
-            (dependencyMap.get(dependent) || new Set()).add(a),
-          )
-          dirtyMap.set(dependent, (dirtyMap.get(dependent) || 0) + 1)
-          loop1(dependent)
-        }
-      })
+
+    // This is a topological sort via depth-first search, slightly modified from
+    // what's described here for performance reasons:
+    // https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search
+
+    // Step 1: find all atoms in the graph
+    const allAtomsInDependencyGraph = new Array<AnyAtom>()
+    const seen = new Set<AnyAtom>()
+    const queue = [atom]
+    while (true) {
+      const a = queue.pop()
+      if (!a) {
+        break
+      }
+      if (!seen.has(a)) {
+        seen.add(a)
+        allAtomsInDependencyGraph.push(a)
+        queue.push(...getDependents(a))
+      }
     }
-    loop1(atom)
-    const loop2 = (a: AnyAtom) => {
-      getDependents(a).forEach((dependent) => {
-        if (dependent !== a) {
-          let dirtyCount = dirtyMap.get(dependent)
-          if (dirtyCount) {
-            dirtyMap.set(dependent, --dirtyCount)
-          }
-          if (!dirtyCount) {
-            let isChanged = !!dependencyMap.get(dependent)?.size
-            if (isChanged) {
-              const prevAtomState = getAtomState(dependent)
-              const nextAtomState = readAtomState(dependent, true)
-              isChanged = !isEqualAtomValue(prevAtomState, nextAtomState)
-            }
-            if (!isChanged) {
-              dependencyMap.forEach((s) => s.delete(dependent))
-            }
+
+    // Step 2: traverse the dependency graph to build the topsorted atom list
+    // We don't bother to check for cycles, which simplifies the algorithm.
+    const topsortedAtoms = new Array<AnyAtom>()
+    const markedAtoms = new Set<AnyAtom>()
+    const visit = (n: AnyAtom) => {
+      if (markedAtoms.has(n)) {
+        return
+      }
+      markedAtoms.add(n)
+      for (const m of getDependents(n)) {
+        if (n !== m) {
+          visit(m)
+        }
+      }
+      // The algorithm calls for pushing onto the front of the list. Since
```

Comment:
I think we can create an array with size as we know the max size, and fill it from the front.

Reply:
```ts
const topsortedAtoms = new Array<AnyAtom>(allAtomsInDependencyGraph.length)
```
Is probably a good idea either way, but even then, I still think we want to avoid `unshift` here. I suppose this could keep track of an index an insert it that way:
```
const index = allAtomsInDependencyGraph.length

topsortedAtoms[--index] = n
```
But I'm not sure that is more clear than iterating `topsortedAtoms` in reverse in step 3.

Reply:
I think I wasn't clear about my suggestion.
```js
const topsortedAtoms = new Array<AnyAtom>(allAtomsInDependencyGraph.length)
let topsortedAtomsIndex = 0
topsortedAtoms[topsortedAtomsIndex++] = n
```

Reply:
Good point, done